### DISCUSSION TITLE
println() in a graphical program

### DISCUSSION THREAD

**Discussion by Ben-Fields** [vlang/v#8794] [General]
Importing the 'ui' or 'gx' modules seems to completely disable terminal output. Is this a bug or do I need to explicitly interface with the terminal in some other way when creating a graphical application? In either case, there seems to be non-obvious change to behavior (contradicting the simplicity mantra). Perhaps I am missing something right in front of me?

```
import ui
fn main() {
    println("How does one print this text?")
}
```

Run from cmd and PowerShell on Win10.

**Comment by JalonSolov** [comment]
It's a Windows thing.  Add `[console]` between your `import` and `fn main` and see what happens.

**Reply by medvednikov** [reply]
Indeed.

Ideally V should detect GUI Windows apps calling println and just make it work :)

**Reply by Ben-Fields** [reply]
Much appreciated. Perhaps a quick addition [to the docs](https://github.com/vlang/v/blob/master/doc/docs.md#attributes).

I've been having fun experimenting with the language.

**Reply by dumblob** [reply]
Hopefully `[console]` is just a temporary solution until it gets removed for *V 1.0* release. Starting with *V 1.0* ConPTY as described in https://github.com/vlang/v/issues/8800#issuecomment-782048461 shall be used everywhere.

**Reply by spytheman** [reply]
Yes, it is a temporary solution. If you know how to use ConPTY @dumblob , a PR will be much appreciated.

**Reply by dumblob** [reply]
@spytheman there is a minimal "ping-echo" app using the new ConPTY API (which consists of **only** 3 functions with very few arguments!) in https://github.com/microsoft/terminal/blob/main/samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp (explanation of core parts in the blog post https://devblogs.microsoft.com/commandline/windows-command-line-introducing-the-windows-pseudo-console-conpty/ ).

**Reply by spytheman** [reply]
Cool. So it will be easy for you to implement it. Thanks in advance.

**Reply by dumblob** [reply]
I'd rather pay someone who knows the *V* code base than to spend hours with just figuring the internals of the *V* compiler myself. I'm much faster in reading small stuff like ConPTY and building a bigger picture for others than spending hours with all the "devil details" of *V* internals without any bigger picture. Is it actually possible to do *V* development on Android smartphones? I'm doing much of my contributions in *V* bug tracker etc. from such constrained devices because I'm often in a hurry from one place to another and not in front of my notebook.

**Comment by tzSharing** [comment]
I have been plagued by this problem recently, because my v ui program will have an "application not responding" problem, and I can only force it to end in the end. The most depressing thing is that the whole process does not output any valuable information. I can only manually check the problem line by line, and finally found that it is because of the use of println (err.msg).
How should it be solved? Glad to have found this discussion article, now that I've solved it. But I think adding `[console]` to the main function confuses me a bit. Because in Android development I use `Log.e()` `Log.w()` `Log.i()` `Log.d()` ... I'm used to it.

**Comment by Wajinn** [comment]
As you have imported ui, it's debatable to use message_box instead of println.  Maybe it could be improved to have message_box with a timer, so it could close itself after a bit.

```
import ui

fn main() {
    showmsg("How does one print this text?")
}


fn showmsg(text string) {
	ui.message_box(text)
}
```

**Reply by ghost** [reply]
Does `message_box()` work on Windows?
Not on Linux.

And currently, `ui` is not mature enough to suggest that using the `ui` module's functionality for error reporting is a good idea.
